input-output-hk/catalyst-voices
# feat(cat-voices): Integration tests using flutter_driver

# Description

The result of the research on how to make an integration test in Flutter web, where extensions need to be installed on the tested browser.

## Related Issue(s)

Resolves #1010

## Description of Changes

- Creating internal package that creates custom driver (VoicesWebDriver)
- Creating simple example how to use it in package catalyst_caradno example.

### How to make tests
1. Make sure that for know you are using the master channel of flutter `flutter channel master`
2. Add packages to pubspec: `catalyst_voices_driver, flutter_driver, webdriver`
3. Create a folder `test_driver` in root directory of project that you want to test
4 Create files in `test_driver` folder: `app_test.dart` and `app.dart` 

### `app.dart` file

```dart
import 'package:catalyst_cardano_example/main.dart' as app;
import 'package:flutter_driver/driver_extension.dart';

void main() {
  enableFlutterDriverExtension();
  app.main();
}
```


### `app_test.dart` file

```dart 
import 'package:catalyst_voices_driver/catalyst_voices_driver.dart';
import 'package:flutter_driver/flutter_driver.dart';
import 'package:test/test.dart';
import 'package:webdriver/async_io.dart';

void main() {
  group('Test if extensions are installed and driver can communicate with them',
      () {
    late final VoicesWebDriver driver;

    // Connect to the Flutter driver before running any tests.
    setUpAll(() async {
      driver = await VoicesWebDriver.create(extensions);
      // Here should be call function to setup wallets
      await driver.switchToWindow('localhost');
    });

    // Close the connection to the driver after the tests have completed.
    tearDownAll(() async {
      await driver.webDriver.quit();
      await driver.close();
    });

    test(
      'Try to enable eternl wallet. Eternl wallet is not configured.',
      () async {
        final buttonFinder = find.byValueKey('enableWallet-eternl');
        await driver.tap(buttonFinder);
        await Future<void>.delayed(const Duration(seconds: 6));
        final currentWindow = await driver.webDriver.window;

        await driver.switchToWindow(extensions[0].id);

        final cancelButtonFinder =
            await driver.webDriver.findElement(const By.tagName('button'));

        await cancelButtonFinder.click();
        await currentWindow.setAsActive();
        find.text('WalletApiException');
      },
    );
  });
}
```

To run test:
1. Run chromdriver in new terminal 
```bash
chromedriver --port=4444
```

2. To run tests
```bash
flutter drive --target=test_driver/app.dart \                                                                                                                                                                                                                        
--web-browser-flag=--disable-web-security \
--web-browser-flag=--disable-gpu \
--web-browser-flag=--disable-search-engine-choice-screen \
--web-header=Cross-Origin-Opener-Policy=same-origin \
--web-header=Cross-Origin-Embedder-Policy=require-corp \
--debug \
--no-headless \
-d web-server --browser-name=chrome --driver-port=4444
```
## Functionality to be implemented

- [ ] downloading extensions before init VoicesWebDriver
- [ ] setting up wallets after init VoicesWebDriver but before tests, inside the `setUpAll` function.
- [ ] handling downloaded extensions (Should they be committed to the project?)

## Breaking Changes

This works only on the master channel of Flutter!

## Please confirm the following checks
